### PR TITLE
Async Lazy_frame.{collect, collect_all, profile, fetch}

### PR DESCRIPTION
**Pull request by somilgo**
I went for `Lazy_frame.Deferred` as the parent module for these functions. Originally, I created a new `Deferred` file and had `Deferred.Lazy_frame`, but I actually like this better organization-wise and it was less code. I tested this in our code and it does indeed parallelize.

### CODE REVIEW

**Comment by somilgo** [comment]
I went for [Lazy_frame.Deferred]. Originally, I create a new [Deferred] module and had [Deferred.Lazy_frame], but I actually like this better organization-wise and it was less code. FWIW, I'm still unable to get rustfmt to work properly in this repo. I tested this in our code and it does indeed parallelize, although I couldn't think of a good test.

**Comment by mt-caret** [comment]
Hmm, while I agree how this is aesthetically quite nice; I'm not sure I completely agree with this approach.

Unfortunately the open-source OCaml ecosystem has a more widely used asynchronous runtime called Lwt,[1] and it's probably not great to strongly couple this library to a particular async runtime.[2] I'm also worried that if we don't explicitly shadow the functions people will accidentally use the non-async versions, which I think is almost certainly not what they want, conditioned on them writing code in async world.

What do you think?

[1]: see http://rgrinberg.com/posts/abandoning-async/ on the async-lwt split
[2]: I think for our particular case, I think it makes sense to split out the async portion into a separate library, similar to how https://github.com/inhabitedtype/angstrom does (see the async and lwt directories).

**Comment by somilgo** [comment]
Ah, Lwt slipped my mind - been a while since I've written non-Core/Async :). I also think making shadowing easier is a compelling argument.

Would you be more in favor of having an async library that sits next to lib and gets imported as Polars_ocaml_async? Then, the expected pattern would be to include both libraries at the top of the file (with the async one second).

Are there other functions you'd like to see async-ified in this PR?

**Comment by mt-caret** [comment]
> Would you be more in favor of having an async library that sits next to lib and gets imported as Polars_ocaml_async? Then, the expected pattern would be to include both libraries at the top of the file (with the async one second).

This sounds reasonable to me!

> Are there other functions you'd like to see async-ified in this PR?

I think ideally we'd probably want to make some of the dataframe operations asynchronous, but I don't think we need to do everything at once (of course, if you're up for doing that, that'll be fantastic!).

File: rust/src/utils.rs [hunk 1]
```diff
@@ -496,3 +496,17 @@ unsafe impl<T: 'static + Clone> ToOCaml<DynBox<T>> for Abstract<T> {
 pub fn unwrap_abstract_vec<T>(v: Vec<Abstract<T>>) -> Vec<T> {
     v.into_iter().map(|Abstract(v)| v).collect()
 }
+
+pub fn maybe_release_runtime<T, F>(cr: &mut OCamlRuntime, should_release_runtime: bool, f: F) -> T
+where
+    F: FnOnce() -> T,
+{
+    if should_release_runtime {
```
Comment (mt-caret) [inline rust/src/utils.rs:504]: Is there a particular reason why we make releasing the runtime lock optional? I would've assumed that for these sorts of blocking operations it's always fine to release the runtime lock.

**Review by mt-caret — CHANGES_REQUESTED** [review]
Can we also add some tests that demonstrate that the async versions of the functions work?

File: rust/src/utils.rs (hunk 1 above)
Comment (somilgo) [reply rust/src/utils.rs:504]: Hmm yeah I wasn't too sure about this either. One failure mode that I could think of is if someone wraps the non-Async version in an In_thread.run (because it is in some larger function with other parallelizable bits), this function might unexpectedly run concurrently with other code. The user might end up with more concurrency than they'd like or maybe they have some logic to cache Data_frames to avoid duplicate work that gets ignored since the code no longer has the lock. 

What do you think? Am I overly cautious or off-base somewhere?

**Comment by somilgo** [comment]
Yeah happy to add some tests that show correctness. Do you have any ideas for a deterministic test that would prove concurrency?

**Comment by mt-caret** [comment]
> Yeah happy to add some tests that show correctness. Do you have any ideas for a deterministic test that would prove concurrency?

Ah, I was thinking of much more basic tests that e.g. are just copies of the existing expect tests in the guide where we call the async versions so it's obvious that the functions don't crash, for example. I'm actually not quite sure how the current code can result in concurrency issues, do you have particular edge cases in mind?

**Comment by somilgo** [comment]
Cool, happy to add those tests that are similar to the existing ones.

I was just thinking something that proves that we're actually releasing the run time lock such that multiple collects can happen at the same time. But I'm fine to just not have that.

File: rust/src/utils.rs (hunk 1 above)
Comment (mt-caret) [reply rust/src/utils.rs:504]: > One failure mode that I could think of is if someone wraps the non-Async version in an In_thread.run (because it is in some larger function with other parallelizable bits), this function might unexpectedly run concurrently with other code.
Ah, I hadn't considered this case; that's a fair point. I see that upstream hasn't put a lot of thought into this problem (judging from https://github.com/pola-rs/polars/issues/8288), but at the same time if there's enough work to be done such that there's a ton of contention in Polar's thread pool, *not* releasing the runtime lock in an Async context to me seems like we'll end up in a worse situation (long async cycles, RPC timeouts, etc.).

> or maybe they have some logic to cache Data_frames to avoid duplicate work that gets ignored since the code no longer has the lock.
I'm not sure I follow; could you elaborate a bit on this case?

I think it might be worth leaving a comment in the code highlighting this possibility, but in the meantime it doesn't seem crazy to me to always release the runtime lock (or, create a lazy_static variable (or thread_local? I'm not sure how that would work with thread pools in async land) that toggles this behavior and defaults to always release the runtime lock). I'm guessing there will eventually be a lot of async functions exposed and if we need to pass a boolean variable around that seems super verbose and cumbersome, even if they are optional arguments.

**Comment by mt-caret** [comment]
> I was just thinking something that proves that we're actually releasing the run time lock such that multiple collects can happen at the same time. But I'm fine to just not have that.

I'm sorry, I'm not sure I fully follow here; what do we get by making sure that multiple collects happen at the same time?

**Comment by somilgo** [comment]
Created new PR #55 since there were substantial changes.
